binder/src/main/java/io/grpc/binder/ServerSecurityPolicy.java

```diff
@@ -44,16 +46,31 @@ private ServerSecurityPolicy(ImmutableMap<String, SecurityPolicy> perServicePoli
   /**
    * Return whether the given Android UID is authorized to access a particular service.
    *
-   * <b>IMPORTANT</b>: This method may block for extended periods of time.
+   * <p><b>IMPORTANT</b>: This method may block for extended periods of time.
    *
    * @param uid The Android UID to authenticate.
    * @param serviceName The name of the gRPC service being called.
+   * @deprecated Application code should not need to call this method.
    */
   @CheckReturnValue
+  @Deprecated
   public Status checkAuthorizationForService(int uid, String serviceName) {
     return perServicePolicies.getOrDefault(serviceName, defaultPolicy).checkAuthorization(uid);
   }
 
+  /**
+   * Return whether the given Android UID is authorized to access a particular service.
+   *
+   * @param uid The Android UID to authenticate.
+   * @param serviceName The name of the gRPC service being called.
+   */
+  @CheckReturnValue
+  ListenableFuture<Status> checkAuthorizationForServiceAsync(int uid, String serviceName) {
+    SecurityPolicy securityPolicy = perServicePolicies.getOrDefault(serviceName, defaultPolicy);
+    return Futures.immediateFuture(securityPolicy.checkAuthorization(uid));
```

Comment:
I think creating a default implementation that blocks is a dangerous short cut. Callers of a method with async in the name that returns a future are not going to expect blocking. Instead I think we should create a new interface called something like AsyncServerSecurityPolicy and support old blocking ServerSecurityPolicies simply by submit()ing them to a ListeningExecutorService (gRPC offload executor?)

Reply:
If for some reason we really need a single ServerSecurityPolicy interface to do everything, perhaps the default checkAuthorizationForServiceAsync() could return null or some sentinel value to indicate this function isn't supported. 

Or perhaps AsyncServerSecurityPolicy extends ServerSecurityPolicy and adds a default implementation of checkAuthorizationForService() that calls get() on the future returned by checkAuthorizationForServiceAsync(). Internally, we could use instanceof to test for AsyncServerSecurityPolicy and prefer that one if available.

Just brainstorming here. Both of these are pretty ugly  ...

Reply:
See my comment above. There's an `AsyncSecurityPolicy` coming soon ([sneak peak](https://github.com/mateusazis/grpc-java/commit/9a94b59bd4ae1ce55ba38cb9288d267b0ae1cd12)) but it was moved out to make each PR smaller.

Please note that:
- `ServerSecurityPolicy.checkAuthorizationForServiceAsync` currently **does not** block per se; just wraps the original status in a future. So I expect it to be just as performant as the current implementation.
- `BinderTransportSecurity.TransportAuthorizationState.checkAuthorization` **does** block, but, for the time being, should display equivalent performance to the current implementation since async policies are not supported yet (i.e. slow `SecurityPolicy` implementions would block anyway). That method has a TODO which I can't address until I add a few more classes, which I'm holding for now to keep this PR small.

Reply:
OK. I get that you are breaking this up into multiple PRs to make each one smaller. And I agree that what you have is no worse performance wise or blocking-wise than what we have today. But I think it's important for each of these small PRs to leave the code in a rational and readable intermediate state. I don't see how that can be done without AsyncSecurityPolicy but I see that's something you've already negotiated with markb74 so maybe I'll just get out of the way.

Reply:
FWIW, what I didn't like about adding AsyncSecurityPolicy in this PR is that it exposed an API which doesn't do what you'd expect.

This state, while a little ugly, is at least internal only.

In any case, I think we can hold off on merging this until we're ready with the follow up.